In [226]:
import pandas as pd

In [227]:
df_movies=pd.read_csv("data/IMDb names.csv")

##lets see how to import data from CSV, API(JSON), SQL Database....etc and hold the data in a dataframe object

##using head() method to get the first few rows
##Note:
##head() - fuction used to get the first n rows
##tail() - function used to get the last n rows

In [228]:
df_movies.head()

imdb_name_id             name               birth_name  height  \
0    nm0000001     Fred Astaire  Frederic Austerlitz Jr.   177.0   
1    nm0000002    Lauren Bacall        Betty Joan Perske   174.0   
2    nm0000003  Brigitte Bardot          Brigitte Bardot   166.0   
3    nm0000004     John Belushi        John Adam Belushi   173.0   
4    nm0000005   Ingmar Bergman     Ernst Ingmar Bergman   179.0   

                                                 bio  \
0  Fred Astaire was born in Omaha, Nebraska, to J...   
1  Lauren Bacall was born Betty Joan Perske on Se...   
2  Brigitte Bardot was born on September 28, 1934...   
3  John Belushi was born in Chicago, Illinois, US...   
4  Ernst Ingmar Bergman was born July 14, 1918, t...   

                                       birth_details  birth_year  \
0               May 10, 1899 in Omaha, Nebraska, USA      1899.0   
1  September 16, 1924 in The Bronx, New York City...      1924.0   
2                September 28, 1934 in Paris, France      1934.0   
3         January 24, 1949 in Chicago, Illinois, USA      1949.0   
4      July 14, 1918 in Uppsala, Uppsala län, Sweden      1918.0   

  date_of_birth                           place_of_birth  \
0    1899-05-10                     Omaha, Nebraska, USA   
1    1924-09-16  The Bronx, New York City, New York, USA   
2    1934-09-28                            Paris, France   
3    1949-01-24                   Chicago, Illinois, USA   
4    1918-07-14             Uppsala, Uppsala län, Sweden   

                                       death_details  death_year  \
0  June 22, 1987 in Los Angeles, California, USA ...      1987.0   
1  August 12, 2014 in New York City, New York, US...      2014.0   
2                                                NaN         NaN   
3  March 5, 1982 in Hollywood, Los Angeles, Calif...      1982.0   
4        July 30, 2007 in Fårö, Gotlands län, Sweden      2007.0   

  date_of_death                             place_of_death  \
0    1987-06-22             Los Angeles, California, USA     
1    2014-08-12             New York City, New York, USA     
2           NaN                                        NaN   
3    1982-03-05  Hollywood, Los Angeles, California, USA     
4    2007-07-30                 Fårö, Gotlands län, Sweden   

                         reason_of_death  spouses  divorces  \
0                              pneumonia        2         0   
1                                 stroke        2         1   
2                                    NaN        4         3   
3  acute cocaine and heroin intoxication        1         0   
4                                    NaN        5         4   

   spouses_with_children  children              primary_profession  \
0                      1         2  soundtrack,actor,miscellaneous   
1                      2         3              actress,soundtrack   
2                      1         1     actress,soundtrack,producer   
3                      0         0         actor,writer,soundtrack   
4                      5         8           writer,director,actor   

                          known_for_titles  
0  tt0050419,tt0053137,tt0072308,tt0043044  
1  tt0037382,tt0038355,tt0117057,tt0071877  
2  tt0054452,tt0059956,tt0057345,tt0049189  
3  tt0078723,tt0072562,tt0080455,tt0077975  
4  tt0050976,tt0083922,tt0069467,tt0050986

In [229]:
df_movies.loc[4]

imdb_name_id                                                     nm0000005
name                                                        Ingmar Bergman
birth_name                                            Ernst Ingmar Bergman
height                                                                 179
bio                      Ernst Ingmar Bergman was born July 14, 1918, t...
birth_details                July 14, 1918 in Uppsala, Uppsala län, Sweden
birth_year                                                            1918
date_of_birth                                                   1918-07-14
place_of_birth                                Uppsala, Uppsala län, Sweden
death_details                  July 30, 2007 in Fårö, Gotlands län, Sweden
death_year                                                            2007
date_of_death                                                   2007-07-30
place_of_death                                  Fårö, Gotlands län, Sweden
reason_of_death          

##import new csv creating an index this time around using column imd_titel.id will no longer be a be acolumn in our data

In [230]:
df_movies = pd.read_csv("data/IMDb movies.csv",index_col ="imdb_title_id")

In [231]:
df_movies.head(2)

title               original_title  year  \
imdb_title_id                                                                   
tt0000574      The Story of the Kelly Gang  The Story of the Kelly Gang  1906   
tt0001892                   Den sorte drøm               Den sorte drøm  1911   

              date_published                    genre  duration  \
imdb_title_id                                                     
tt0000574         1906-12-26  Biography, Crime, Drama        70   
tt0001892         1911-08-19                    Drama        53   

                        country language      director  \
imdb_title_id                                            
tt0000574             Australia      NaN  Charles Tait   
tt0001892      Germany, Denmark      NaN     Urban Gad   

                                              writer  ...  \
imdb_title_id                                         ...   
tt0000574                               Charles Tait  ...   
tt0001892      Urban Gad, Gebhard Schätzler-Perasini  ...   

                                                          actors  \
imdb_title_id                                                      
tt0000574      Elizabeth Tait, John Tait, Norman Campbell, Be...   
tt0001892      Asta Nielsen, Valdemar Psilander, Gunnar Helse...   

                                                     description avg_vote  \
imdb_title_id                                                               
tt0000574      True story of notorious Australian outlaw Ned ...      6.1   
tt0001892      Two men of high rank are both wooing the beaut...      5.9   

               votes  budget usa_gross_income worlwide_gross_income metascore  \
imdb_title_id                                                                   
tt0000574        537  $ 2250              NaN                   NaN       NaN   
tt0001892        171     NaN              NaN                   NaN       NaN   

               reviews_from_users  reviews_from_critics  
imdb_title_id                                            
tt0000574                     7.0                   7.0  
tt0001892                     4.0                   2.0  

[2 rows x 21 columns]

In [232]:
df_test = pd.read_csv("data/IMDb movies.csv")

## Lets confirm if the imdb_title _id is really unique for every rowusing the count function in pandas and storing in a set

In [233]:
x= df_test["imdb_title_id"]

In [234]:
x.count()

81273

In [235]:
len(set())

0

###importing Data from REST API (i.eJSON) json is a dattatype of different languages
in python we decode JSON into dictionaries

In [236]:
df_posts=pd.read_json("https://jsonplaceholder.typicode.com/posts")

In [237]:
df_posts["id"].count()==len(set(df_posts["id"]))

True

In [238]:
df_posts.head(1)

userId  id                                              title  \
0       1   1  sunt aut facere repellat provident occaecati e...   

                                                body  
0  quia et suscipit\nsuscipit recusandae consequu...

In [239]:
df_posts.set_index("id", inplace=True)

In [240]:
df_posts.head(2)

userId                                              title  \
id                                                              
1        1  sunt aut facere repellat provident occaecati e...   
2        1                                       qui est esse   

                                                 body  
id                                                     
1   quia et suscipit\nsuscipit recusandae consequu...  
2   est rerum tempore vitae\nsequi sint nihil repr...

##Getting data from and sql database

In [241]:
!pip install Mysql -python


Usage:   
  pip install [options] <requirement specifier> [package-index-options] ...
  pip install [options] -r <requirements file> [package-index-options] ...
  pip install [options] [-e] <vcs project url> ...
  pip install [options] [-e] <local project path> ...
  pip install [options] <archive url/path> ...

no such option: -p


In [242]:
!pip install ps


In [243]:
!conda install pscopg2-binary

Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.
Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.



PackagesNotFoundError: The following packages are not available from current channels:

  - pscopg2-binary

Current channels:

  - https://repo.anaconda.com/pkgs/main/win-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/win-64
  - https://repo.anaconda.com/pkgs/r/noarch
  - https://repo.anaconda.com/pkgs/msys2/win-64
  - https://repo.anaconda.com/pkgs/msys2/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.




In [244]:
import psycopg2

con = ps.connect("host=localhost user=postgres password=12345 dbname=data_sales")

df = pd.read_sql_query("SELECT * FROM fruits", con)

df

NameError: name 'ps' is not defined

In [ ]:
df_movies.head(2)

### prerequisites of EDA(Exploratory Data Analysis)i.e descriptive statistics

### Data Preparation and cleaning steps are;

* run df.info()to check column /row frequencies of null/non- null values. 
* If more than half of the rows in a column has none/NA/Null values then we could drop the column
* If the null values are few, then we can replace them with mean/median/mode
* Deal with duplicate values. Remove and keep only one.


### Performing descriptive statistics

* Also run describe() method on your df. This will show frequency /mean/min/max/standard deviation/variance of continuous values(numbers only).
* Check the relationship between continous variables i.e one column and another


### Data Visualization (charts)


In [ ]:
df_movies.info()

## country seams to have 39 null values --? How would we deal with this?

we replace the values with a measure of tendency since they are few, In this case we choose the mode (most common country) because the data type is object(string).

_NB : If say we needed the actors column in our objectives then we would have to remove all rows which do not have actors.

_This would be the case if you are performing a qualitative anaylsis instead of quantitative

## Budget column has 22804 out of 81273 (i.e almost 72% null values. How would we deal with the scenario?

we would drop the entire column since it contains too many null values

### How would we deal with the column reviews_from_users?

We replace the null values with the mean. The data type is integer.

## An example of creating dataframe from sql

In [ ]:
!pip install psycopg2

In [ ]:
import psycopg2 as ps

con = ps.connect("host=localhost user=postgres password=12345 dbname=data_sales")

df = pd.read_sql_query("SELECT * FROM fruits", con)

df

## Clean up our columns and check for duplicates

In [ ]:
df_movies.head(1)

In [ ]:
df_movies.info()

In [ ]:
df_movies.head()

In [ ]:
df_movies_one=pd.read_csv("data/imdb_data_1000.csv", index_col='Rank')

In [ ]:
df_movies_one.head(1)

In [ ]:
df_movies_one["Rank"].count()==len(set(df_movies_one["Rank"]))

In [ ]:
df_movies_one.rename(columns ={'Runtime (Minutes)':'Runtime_minutes', 'revenue (millions)':'revenue_millions'}, inplace =True)

In [ ]:
df_movies_one.columns

In [ ]:
df_movies_one.columns = [c.lower() for c in df_movies_one]

In [ ]:
df_movies_one.columns

In [ ]:
df_movies_one.shape

In [ ]:
df_movies_one.drop_duplicates(keep='first', inplace = True)

In [ ]:
df_movies_one.shape

##Create a temporary dataframe adn create duplicate values using append()method

In [ ]:
tmp_df=df_movies_one.append(df_movies_one)

In [ ]:
tmp_df.shape

In [ ]:
tmp_df=tmp_df.append(df_movies_one)

In [ ]:
tmp_df.shape

In [ ]:
tmp_df_one=tmp_df.copy()

In [ ]:
tmp_df.drop_duplicates(inplace=True,keep='last')

In [ ]:
tmp_df.shape

In [ ]:
tmp_df_one.drop_duplicates(inplace=True, keep=False)

In [ ]:
tmp_df_one.shape


In [ ]:
df_movies_one.head()

In [ ]:
df_movies.shape

In [ ]:
df_movies_one.shape

In [ ]:
df_comb = df_movies_one.append(df_movies)

In [ ]:
df_comb.shape

In [ ]:
df_comb.columns

In [ ]:
df_movies.columns

In [ ]:
df_movies_one.columns

### Dealing with Null values

In [ ]:
df_movies_one.shape

In [ ]:
df_movies.shape

In [ ]:
df_nulls=df_movies_one.isnull()

In [249]:
df_nulls.sum()

title                   0
genre                   0
description             0
director                0
actors                  0
year                    0
runtime_minutes         0
rating                  0
votes                   0
revenue (millions)    128
metascore              64
dtype: int64

### Removes any row which has atleast 1 null value

In [ ]:
df_movies_one1= df_movies_one.dropna()

In [ ]:
df_movies_one1.shape

### Alternative way of removing rows with null values

In [ ]:
df_movies_one3=df_movies_one.copy()

In [ ]:
df_movies_one3.dropna(axis=0, inplace=True)

In [ ]:
df_movies_one3.shape

In [ ]:
df_movies_one2=df_movies_one.copy()

### Remove a column with at least one null value

In [ ]:
df_movies_one2.dropna(axis=1, inplace=True)

In [ ]:
df_movies_one2.shape

In [263]:
revenue = df_movies_one['revenue_millions']

In [264]:
revenue.head()

Rank
1    333.13
2    126.46
3    138.12
4    270.32
5    325.02
Name: revenue_millions, dtype: float64

In [265]:
type(revenue)

pandas.core.series.Series

In [276]:
revenuemean=revenue.mean()

In [277]:
revenuemean

82.9563761467888

In [278]:
revenue.fillna(revenuemean, inplace =True)

In [279]:
revenue.shape

(1000,)

### Find how many values in series have the value "revenuemean" as above

In [281]:
df_movies_one.isnull().sum()

title                0
genre                0
description          0
director             0
actors               0
year                 0
runtime_minutes      0
rating               0
votes                0
revenue_millions     0
metascore           64
dtype: int64

In [282]:
d5=df_movies_one.copy()

In [287]:
d5['metascore'].fillna(d5['metascore'].mean(),inplace=True)

In [288]:
d5.isnull().sum()

title               0
genre               0
description         0
director            0
actors              0
year                0
runtime_minutes     0
rating              0
votes               0
revenue_millions    0
metascore           0
dtype: int64

In [291]:
d5.to_csv('data/imdb_1000_clean.csv')